apps/opik-python-backend/src/opik_backend/executor_docker.py

```diff
@@ -0,0 +1,137 @@
+import atexit
+import concurrent.futures
+import json
+import logging
+import os
+from queue import Queue
+from threading import Lock
+from uuid6 import uuid7
+
+import docker
+
+from opik_backend.executor import CodeExecutorBase, ExecutionResult
+from opik_backend.scoring_commands import PYTHON_SCORING_COMMAND
+
+logger = logging.getLogger(__name__)
+client = docker.from_env()
+executor = concurrent.futures.ThreadPoolExecutor()
+
+class DockerExecutor(CodeExecutorBase):
+    def __init__(self):
+        super().__init__()
+        # Docker-specific configuration
+        self.docker_registry = os.getenv("PYTHON_CODE_EXECUTOR_IMAGE_REGISTRY", "ghcr.io/comet-ml/opik")
+        self.docker_image = os.getenv("PYTHON_CODE_EXECUTOR_IMAGE_NAME", "opik-sandbox-executor-python")
+        self.docker_tag = os.getenv("PYTHON_CODE_EXECUTOR_IMAGE_TAG", "latest")
+
+        self.instance_id = str(uuid7())
+        self.container_labels = {"managed_by": self.instance_id}
+        self.container_pool = Queue()
+        self.pool_lock = Lock()
+        self.ensure_pool_filled()
+        atexit.register(self.cleanup)
+
+    def ensure_pool_filled(self):
+        with self.pool_lock:
+            while len(self.get_managed_containers()) < self.max_parallel:
+                logger.warning("Not enough containers running; creating more...")
+                self.create_container()
+
+    def get_managed_containers(self):
+        return client.containers.list(filters={
+            "label": f"managed_by={self.instance_id}",
+            "status": "running"
+        })
+
+    def create_container(self):
+        new_container = client.containers.run(
+            image=f"{self.docker_registry}/{self.docker_image}:{self.docker_tag}",
+            command=["tail", "-f", "/dev/null"],
+            mem_limit="128mb",
+            cpu_shares=2,
+            detach=True,
+            network_disabled=True,
+            security_opt=["no-new-privileges"],
+            labels=self.container_labels
+        )
+        self.container_pool.put(new_container)
+
+    def release_container(self, container):
+        def async_release():
+            try:
+                logger.debug(f"Stopping container {container.id}. Will create a new one.")
+                container.stop(timeout=1)
+                container.remove(force=True)
+                self.create_container()
+            except Exception as e:
+                logger.error(f"Error replacing container: {e}")
+
+        executor.submit(async_release)
+
+    def get_container(self):
+        while True:
```

Comment:
What about adding a Circuit breaker flag, which is disabled for instance during application shutdown